File: 871.py
# 871. Minimum Number of Refueling Stops
from typing import List

class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        n_stations = len(stations)
        dp = [0] * (n_stations + 1)
        dp[0] = startFuel

        for i in range(n_stations):
            for t in range(i, -1, -1):
                if dp[t] >= stations[i][0]:
                    dp[t + 1] = max(dp[t + 1], dp[t] + stations[i][1])

        for i in range(n_stations + 1):
            if (dp[i] >= target): return i


        return -1

File: 871.cpp
// 871. Minimum Number of Refuling Stops
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>>& stations) {
        int n_stations = stations.size();
        vector<long> dp (n_stations + 1);
        dp[0] = startFuel;

        for (int i = 0; i < n_stations; i++){
            for (int t = i; t >= 0; t--){
                if (dp[t] >= stations[i][0])
                    dp[t + 1] = max(dp[t + 1], dp[t] + stations[i][1]);
            }
        }

        for (int i = 0; i <= n_stations; i++)
            if (dp[i] >= target) return i;
        
        return -1;
    }
};


int main(){

    return 0;
}